crates/rome_js_analyze/src/semantic_analyzers/nursery/no_confusing_arrow.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{semantic_services::Semantic, JsRuleAction};
+use rome_analyze::{context::RuleContext, declare_rule, ActionCategory, Rule, RuleDiagnostic};
+use rome_console::markup;
+use rome_diagnostics::Applicability;
+use rome_js_factory::make;
+use rome_js_syntax::{
+    AnyJsExpression, AnyJsFunctionBody, JsArrowFunctionExpression, JsSyntaxKind, JsSyntaxToken,
+};
+use rome_rowan::BatchMutationExt;
+
+declare_rule! {
+    /// Disallow arrow functions where they could be confused with comparisons.
+    ///
+    /// Source: https://eslint.org/docs/latest/rules/no-confusing-arrow
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// var x = a => 1 ? 2 : 3;
+    /// var x = (a) => 1 ? 2 : 3;
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// var x = a => (1 ? 2 : 3);
+    /// var x = (a) => (1 ? 2 : 3);
+    /// var x = (a) => {
+    ///     return 1 ? 2 : 3;
+    /// };
+    /// var x = a => { return 1 ? 2 : 3; };
+    /// ```
+    ///
+    pub(crate) NoConfusingArrow {
+        version: "next",
+        name: "noConfusingArrow",
+        recommended: false,
+    }
+}
+
+impl Rule for NoConfusingArrow {
+    type Query = Semantic<JsArrowFunctionExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let arrow_fn = ctx.query();
+
+        if arrow_fn.fat_arrow_token().is_err() {
+            return None;
+        }
+
+        match arrow_fn.body() {
+            Ok(AnyJsFunctionBody::AnyJsExpression(AnyJsExpression::JsConditionalExpression(_))) => {
+                Some(())
+            }
+            _ => None,
+        }
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _state: &Self::State) -> Option<RuleDiagnostic> {
+        let arrow_fn = ctx.query();
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                arrow_fn
+                    .fat_arrow_token()
+                    .expect("We checked that a `fat_arrow_token` exists in `run`")
```

Comment:
Now I see why you make this check in `run`.
In other rules, it is more common to turn the error into an option using `ok()`, and then to use `?` to return if there is an error.

```suggestion
                    .ok()?
```